### DISCUSSION TITLE
[Task]: Add U-Flow model

### DISCUSSION THREAD

**Discussion by mtailanian** [open-edge-platform/anomalib#1367] [Feature Requests]
### What is the motivation for this task?

Hello guys! 

I've been working on a method for Anomaly Detection, which we call U-Flow. It's basically a Normalizing Flows-based method that aggregates the scales' information in UNet-like architecture. 
Also, we included a last stage that computes an automatic threshold for anomaly segmentation, which is based only on statistical hypothesis testing and doesn't make use of any anomalous samples.

This work is currently under review
I included the references in case you want to take a look:
- https://arxiv.org/pdf/2211.12353.pdf 
- https://github.com/mtailanian/uflow

### Describe the solution you'd like

I wonder if you find it of interest that I contribute with this method. I'd be happy to do it.

Thanks in advance, and thank you very much for such a nice work here!

**Comment by samet-akcay** [comment]
Hi @mtailanian, first of all, congrats on your submission, hope you'll get the paper published.

If you would like to contribute, we would be happy to see U-Flow implemented in anomalib. Thanks!

**Comment by mtailanian** [comment]
Thanks @samet-akcay ! I'll start working on this!

**Comment by mtailanian** [comment]
Hello again @samet-akcay
I already implemented the basic functionalities, but I have a doubt about how to include the segmentation method.

Let me add some context: As a final stage of U-Flow, we propose a method for segmenting the anomalies, using an _a contrario_ approach. Basically, it performs a series of statistical tests and decides for each pixel if it is anomalous or not, taking as input the output of the Normalizing Flow graph. 
This part of the method is slower than the rest, and should only be called if a segmentation mask is required.

I'm trying to understand what's the best way of including it, but I'd like to discuss it to get some feedback from this community!

(I can start a PR if you think that looking at the code may help)

Thanks in advance!

**Comment by samet-akcay** [comment]
Since U-Flow has been implemented, I'm now closing this discussion.

**Comment by mtailanian** [comment]
Hello @samet-akcay , hope everything is well.
Regarding U-Flow: the process was long, but the paper got accepted in a journal and is about to be published. 
I'd like to add the last changes, so everything is properly updated.
Is it ok if I open a new pull request adding it?
Thanks!

**Reply by samet-akcay** [reply]
Hi @mtailanian, congratulations! Sure go ahead and create a PR! Thanks a lot for the effort!
